Random123 as a submodule

olupton opened the pull request:
- Update Random123 to the latest tagged release on GitHub (https://github.com/DEShawResearch/random123/releases/tag/v1.14.0), drop the bundled Random123 headers from the NEURON repository.
- Pass `-notatty` to some tests launched via `special`, as these can otherwise hang at an interactive prompt if `ctest` is run from an interactive terminal.
- Disable `coreneuron_standalone::test_nrn_corenrn_standalone` test in GPU builds.
- Update `src/nrniv/geometry3d.cpp` to be compilable with `nvc++ -cuda` by dropping a global-namespace definition of `double max(double, double)`

For sanity, it would be nice to:

- [x] Run modeldb-ci with this PR and previous master: https://github.com/neuronsimulator/nrn-modeldb-ci/actions/runs/2238239513 triggered with `neuron-nightly` and https://dev.azure.com/neuronsimulator/aa1fb98d-a914-45c3-a215-5e5ef1bd7687/_apis/build/builds/5547/artifacts?artifactName=drop&api-version=7.0&%24format=zip
- [x] Trigger PR on coreneuron side to make sure BBP sim stack (and hence Random123) usage is not affecting anything: ~https://bbpgitlab.epfl.ch/hpc/coreneuron/-/pipelines/51307~ https://bbpgitlab.epfl.ch/hpc/coreneuron/-/pipelines/51326
- [x] Check what happens to source tarballs i.e. CMake should be updated with a meaningful error message?

pramodk reviewed (commented):
@olupton: I added two Todos in the description and I think it's worth doing that considering the extent of the changes?

alexsavulescu reviewed (approved):
Given all else turns out OK -> 🚀

olupton commented:
> I think it's worth doing that considering the extent of the changes?

Yep! Both launched now (links in description). I was waiting for the wheel jobs to finish.

olupton commented:
Rebased again and added another local fix I had failed to add: 1c07164f08a16616e242e0547c32085d5f9dbc42.

This fixes compilation with `nvc++ -cuda`; which is not something we do as standard, but it's helpful if it works... (`-cuda` implicitly defines `max`).

olupton commented:
The ModelDB CI and CoreNEURON pipelines both passed too. The links are in the description.

pramodk commented:
@nrnhines : just to be sure, do you think this is all good to go?

@alexsavulescu @olupton : One thing I wanted to mention about the source distribution:

![image](https://user-images.githubusercontent.com/666852/165742083-c26224e7-ffe5-4a49-8caf-b8fdd04374cf.png)

Until now default source tarballs were **sufficient (only) for basic neuron installation i.e. vanilla neuron** without coreneuron without interviews etc. This is why we started using `full-src-package-x.y.0.tar.gz`.

But after this PR, I assume if someone now download tarball then CMake will fail without any useful error message. So at least we should check this and make sure CMake gives useful error message (`Hey, source tarballs are incomplete, either download full-src... or git clone`)

alexsavulescu commented:
The GH zips are flawed by default (IIRC). 

> Until now default source tarballs were **sufficient (only) for basic neuron installation i.e. vanilla neuron**

for `nrn-8.0.2.zip`:
```cmake
(venv) bb-c02gx3mmq6lt(arm64): bld savulesc$cmake ..
-- Found Cython: /Users/savulesc/Downloads/nrn-8.0.2/bld/venv/bin/cython (found version "0.29.28") 
-- Found MPI_C: /opt/homebrew/Cellar/open-mpi/4.1.2/lib/libmpi.dylib (found version "3.1") 
-- Found MPI_CXX: /opt/homebrew/Cellar/open-mpi/4.1.2/lib/libmpi.dylib (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
CMake Error at cmake/ExternalProjectHelper.cmake:38 (message):
  Looks like you are building from source.  Git needed for iv feature.
Call Stack (most recent call first):
  CMakeLists.txt:325 (add_external_project)


-- Configuring incomplete, errors occurred!
See also "/Users/savulesc/Downloads/nrn-8.0.2/bld/CMakeFiles/CMakeOutput.log".
See also "/Users/savulesc/Downloads/nrn-8.0.2/bld/CMakeFiles/CMakeError.log"
```

nrnhines reviewed (approved):
Overall, this is consistent with our trend toward submoduleing all our external prerequisites.
Looking forward to the day when I can do the same for sundials integrators.
I take it that the specific machine "features" files we added to R123 (and for machines that still are in use) are now handled within the external repository.

nrnhines commented:
I actually don't care anymore about source tarballs.  I wonder if, for anyone who actually needs one, they could just download a script which creates the tarball. (I don't remember the context of the last need for one.) Since cmake is a critical part of the build process when starting with a tarball, seems like ```ninja tarball``` would suffice. (like I say, I can't remember what the circumstances were that someone could not use git but could download a zip file or tarball.)

pramodk commented:
thanks Michael for the feedback.

@olupton: As we discussed via slack, if you could do minor modifications in the CMakeLists.txt so that there could be meaningful error (in case of non-git tarball) then it will be helpful. Rest is all good to go!

olupton commented:
With the latest changes, trying to run CMake from one of the standard GitHub tarballs gives:
```
olupton@ewaewa build % cmake .. -G Ninja
-- git describe failed (stdout: '', stderr: 'fatal: not a git repository (or any of the parent directories): .git', code: 128), skipping the version check
CMake Error at cmake/ExternalProjectHelper.cmake:15 (message):
  To build NEURON from source you either need to clone the NEURON Git
  repository or download a source code archive that includes Git submodules,
  such as the full-src-package-X.Y.Z.tar.gz file in a NEURON release on
  GitHub (3rdparty.cmake could not be found).
Call Stack (most recent call first):
  CMakeLists.txt:227 (nrn_submodule_file_not_found)
```

pramodk reviewed (approved):
LGTM (I added explicit note about this in 9be8da41261aa8e33ea89b1be9ce1cbaaedfe1c9)
